### PR TITLE
Support for form_post in OIDC responses

### LINKED ISSUE #9319: support for `form_post` response mode for OAuth2
https://openid.net/specs/oauth-v2-form-post-response-mode-1_0.html defines a `form_post` response mode for returning Authorization Response parameters from the Authorization Endpoint. This mode is required by Apple when requesting non-default scopes (https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms), which we believe is required to enable FaceID/TouchID.

I think that supporting `form_post` is pretty trivial; we should do so.

### PR DESCRIPTION
**Pull request by richvdh**
Apple want to POST the OIDC auth response back to us rather than using query-params; add the necessary support to make that work.

Fixes #9319.

### CODE REVIEW

File: synapse/handlers/oidc_handler.py [hunk 1]
```diff
@@ -149,26 +150,29 @@ async def handle_oidc_callback(self, request: SynapseRequest) -> None:
...
-        )
+        # Removing the cookies early avoids spamming the provider with token requests.
+        for cookie_name, options in [
+            (SESSION_COOKIE_NAME, b"; SameSite=None"),
+            (SESSION_COOKIE_NAME_IOS_HACK, b""),
+        ]:
+            request.cookies.append(
+                b"%s=; Path=/_synapse/client/oidc; Expires=Thu, Jan 01 1970 00:00:00 UTC; "
+                b"HttpOnly; Secure%s" % (cookie_name, options)
+            )
```
Comment (clokep) [inline synapse/handlers/oidc_handler.py:175]: This is to get around Twisted not supporting setting `SameSite=None`?

File: synapse/handlers/oidc_handler.py [hunk 2]
```diff
@@ -693,14 +697,33 @@ async def handle_redirect_request(
...
+        #
+        # Secure is necessary for SameSite=None (and, empirically, also breaks things
+        # on iOS 12.)
+        #
+        # we have to build the cookie by hand rather than calling request.addCookie
+        # to work around https://twistedmatrix.com/trac/ticket/10088
+        #
+        for cookie_name, options in [
+            (SESSION_COOKIE_NAME, b"; Secure; SameSite=None"),
+            (SESSION_COOKIE_NAME_IOS_HACK, b""),
```
Comment (clokep) [inline synapse/handlers/oidc_handler.py:721]: These don't seem to match above where we clear out the cookies. In that case *both* are set to `Secure`.

We should probably pull these out to constants.

**Review by clokep — CHANGES_REQUESTED** [review]
Looks OK based on the previous conversations we had about the security aspects of this, I think there's some mismatches though. See the inline comments.

File: synapse/handlers/oidc_handler.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/oidc_handler.py:175]: yes.

File: synapse/handlers/oidc_handler.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/oidc_handler.py:721]: Ooops. fair.

**Review by clokep — APPROVED** [review]
LGTM!

**Pull request merged**